File: packages/provide-uterm-go/serverauth/jwks.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package serverauth

import (
	"crypto/rsa"
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math/big"
	"net/http"
	"sync"
	"time"

	jwt "github.com/golang-jwt/jwt/v5"
)

// jwksCache is the process-wide JWKS document cache keyed by URL, mirroring
// auth._JWKS_CLIENT_CACHE (one entry per issuer in practice, capped).
var (
	jwksCacheMu  sync.Mutex
	jwksCache    = map[string]map[string]any{} // url -> kid -> public key
	jwksCacheMax = 16
	// jwksHTTPClient is the fetcher; overridable in tests.
	jwksHTTPClient = &http.Client{Timeout: 10 * time.Second}
)

type jwkKey struct {
	Kty string `json:"kty"`
	Kid string `json:"kid"`
	N   string `json:"n"`
	E   string `json:"e"`
}

type jwkSet struct {
	Keys []jwkKey `json:"keys"`
}

// resolveJWKSKey fetches (and caches) the JWKS document for url and returns the
// public key matching the token's kid header. Ports the JWKS path of
// _resolve_jwt_key (which Python delegates to PyJWKClient).
func resolveJWKSKey(url string, token *jwt.Token) (any, error) {
	kid, _ := token.Header["kid"].(string)
	jwksCacheMu.Lock()
	keys := jwksCache[url]
	jwksCacheMu.Unlock()
	if keys == nil {
		fetched, err := fetchJWKS(url)
		if err != nil {
			return nil, err
		}
		jwksCacheMu.Lock()
		if len(jwksCache) >= jwksCacheMax {
			for k := range jwksCache { // evict one (map order is arbitrary)
				delete(jwksCache, k)
				break
			}
		}
		jwksCache[url] = fetched
		keys = fetched
		jwksCacheMu.Unlock()
	}
	if key, ok := keys[kid]; ok {
		return key, nil
	}
	// Single-key JWKS documents often omit kid; accept the sole key.
	if len(keys) == 1 {
		for _, key := range keys {
			return key, nil
		}
	}
	return nil, fmt.Errorf("no JWKS key matches kid %q", kid)
}

func fetchJWKS(url string) (map[string]any, error) {
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	resp, err := jwksHTTPClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("JWKS fetch failed: status %d", resp.StatusCode)
	}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var set jwkSet
	if err := json.Unmarshal(body, &set); err != nil {
		return nil, err
	}
	out := map[string]any{}
	for _, k := range set.Keys {
		if k.Kty != "RSA" {
			continue // only RSA JWKs are supported by this port
		}
		pub, err := rsaFromJWK(k)
		if err != nil {
			return nil, err
		}
		out[k.Kid] = pub
	}
	if len(out) == 0 {
		return nil, errors.New("JWKS document contained no usable RSA keys")
	}
	return out, nil
}

func rsaFromJWK(k jwkKey) (*rsa.PublicKey, error) {
	nBytes, err := base64.RawURLEncoding.DecodeString(k.N)
	if err != nil {
		return nil, err
	}
	eBytes, err := base64.RawURLEncoding.DecodeString(k.E)
	if err != nil {
		return nil, err
	}
	// Left-pad the exponent to 8 bytes for uint64 decoding.
	padded := make([]byte, 8)
	copy(padded[8-len(eBytes):], eBytes)
	return &rsa.PublicKey{
		N: new(big.Int).SetBytes(nBytes),
		E: int(binary.BigEndian.Uint64(padded)),
	}, nil
}

// resetJWKSCache clears the cache (test hook).
func resetJWKSCache() {
	jwksCacheMu.Lock()
	jwksCache = map[string]map[string]any{}
	jwksCacheMu.Unlock()
}

File: packages/provide-uterm-csharp/src/Provide.Uterm/ServerAuth/Jwks.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

using System.Security.Cryptography;
using System.Text.Json;
using System.Text.Json.Serialization;
using Microsoft.IdentityModel.Tokens;

namespace Provide.Uterm.ServerAuth;

/// <summary>
/// Process-wide JWKS document cache (url → kid → RSA public key), mirroring Go
/// <c>serverauth/jwks.go</c>: RSA-only, kid match with single-key fallback, max 16 URLs.
/// </summary>
internal static class Jwks
{
    private const int CacheMax = 16;
    private static readonly TimeSpan FetchTimeout = TimeSpan.FromSeconds(10);
    private static readonly object CacheLock = new();
    private static Dictionary<string, Dictionary<string, RsaSecurityKey>> _cache = new(StringComparer.Ordinal);
    private static HttpClient? _httpClient;

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
    };

    /// <summary>Override HTTP client for tests. Pass null to restore the default 10s client.</summary>
    internal static void SetHttpClient(HttpClient? client)
    {
        lock (CacheLock)
        {
            _httpClient = client;
        }
    }

    /// <summary>Clear the process-wide JWKS cache (test hook).</summary>
    internal static void ResetCache()
    {
        lock (CacheLock)
        {
            _cache = new Dictionary<string, Dictionary<string, RsaSecurityKey>>(StringComparer.Ordinal);
        }
    }

    /// <summary>
    /// Fetch (and cache) the JWKS document for <paramref name="url"/> and return the
    /// RSA public key matching <paramref name="kid"/>. Single-key documents accept
    /// tokens with empty/missing kid.
    /// </summary>
    internal static RsaSecurityKey ResolveKey(string url, string? kid)
    {
        if (string.IsNullOrWhiteSpace(url))
        {
            throw new SecurityTokenException("jwt_jwks_url is empty");
        }

        Dictionary<string, RsaSecurityKey>? keys;
        lock (CacheLock)
        {
            _cache.TryGetValue(url, out keys);
        }

        if (keys is null)
        {
            keys = FetchJwks(url);
            lock (CacheLock)
            {
                if (_cache.Count >= CacheMax)
                {
                    // Evict one arbitrary entry (map order is arbitrary).
                    foreach (var existing in _cache.Keys)
                    {
                        _cache.Remove(existing);
                        break;
                    }
                }

                _cache[url] = keys;
            }
        }

        var want = kid ?? "";
        if (keys.TryGetValue(want, out var matched))
        {
            return matched;
        }

        // Single-key JWKS documents often omit kid; accept the sole key.
        if (keys.Count == 1)
        {
            foreach (var sole in keys.Values)
            {
                return sole;
            }
        }

        throw new SecurityTokenException($"no JWKS key matches kid \"{want}\"");
    }

    private static HttpClient Http
    {
        get
        {
            lock (CacheLock)
            {
                return _httpClient ??= CreateDefaultClient();
            }
        }
    }

    private static HttpClient CreateDefaultClient()
    {
        return new HttpClient { Timeout = FetchTimeout };
    }

    private static Dictionary<string, RsaSecurityKey> FetchJwks(string url)
    {
        // Use the async HTTP path so injectable HttpMessageHandler stubs only need SendAsync.
        using var response = Http.GetAsync(url).GetAwaiter().GetResult();
        if (response.StatusCode != System.Net.HttpStatusCode.OK)
        {
            throw new SecurityTokenException($"JWKS fetch failed: status {(int)response.StatusCode}");
        }

        using var stream = response.Content.ReadAsStream();
        var set = JsonSerializer.Deserialize<JwkSet>(stream, JsonOptions)
                  ?? throw new SecurityTokenException("JWKS document was empty or invalid");

        var outKeys = new Dictionary<string, RsaSecurityKey>(StringComparer.Ordinal);
        if (set.Keys is null || set.Keys.Count == 0)
        {
            throw new SecurityTokenException("JWKS document contained no usable RSA keys");
        }

        foreach (var k in set.Keys)
        {
            if (!string.Equals(k.Kty, "RSA", StringComparison.Ordinal))
            {
                continue; // only RSA JWKs are supported
            }

            var pub = RsaFromJwk(k);
            outKeys[k.Kid ?? ""] = pub;
        }

        if (outKeys.Count == 0)
        {
            throw new SecurityTokenException("JWKS document contained no usable RSA keys");
        }

        return outKeys;
    }

    internal static RsaSecurityKey RsaFromJwk(JwkKey k)
    {
        if (string.IsNullOrEmpty(k.N) || string.IsNullOrEmpty(k.E))
        {
            throw new SecurityTokenException("RSA JWK missing n or e");
        }

        byte[] nBytes;
        byte[] eBytes;
        try
        {
            nBytes = Base64UrlEncoder.DecodeBytes(k.N);
            eBytes = Base64UrlEncoder.DecodeBytes(k.E);
        }
        catch (Exception ex)
        {
            throw new SecurityTokenException("invalid base64url in RSA JWK n/e", ex);
        }

        var rsa = RSA.Create();
        try
        {
            rsa.ImportParameters(new RSAParameters
            {
                Modulus = nBytes,
                Exponent = eBytes,
            });
            return new RsaSecurityKey(rsa) { KeyId = k.Kid };
        }
        catch
        {
            rsa.Dispose();
            throw;
        }
    }

    /// <summary>JSON shape for a JWKS document (RSA subset).</summary>
    internal sealed class JwkSet
    {
        [JsonPropertyName("keys")]
        public List<JwkKey>? Keys { get; set; }
    }

    /// <summary>Single JWK entry (RSA fields only).</summary>
    internal sealed class JwkKey
    {
        [JsonPropertyName("kty")]
        public string? Kty { get; set; }

        [JsonPropertyName("kid")]
        public string? Kid { get; set; }

        [JsonPropertyName("n")]
        public string? N { get; set; }

        [JsonPropertyName("e")]
        public string? E { get; set; }
    }
}
